tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1691,6 +1714,90 @@ class LocationSet {
   uint32_t nextLoc[kMaxIndex];              ///< Next available location
 };
 
+/// A class for managing stage input/output packed locations to avoid duplicate
+/// uses of the same location and component.
+class PackedLocationAndComponentSet {
+public:
+  PackedLocationAndComponentSet(SpirvBuilder &spirvBuilder,
+                                llvm::function_ref<uint32_t(uint32_t)> nextLocs)
+      : spvBuilder(spirvBuilder), assignLocs(nextLocs) {}
+
+  bool assignLocAndComponent(const StageVar *var) {
+    if (tryReuseLocations(var)) {
+      return true;
+    }
+    return assignNewLocations(var);
+  }
+
+private:
+  // Try to pack signature of |var| into unused component slots of already
+  // assigned locations. Returns whether it succeeds or not.
+  bool tryReuseLocations(const StageVar *var) {
```

Comment:
I've read over this function a few times, and I'm still struggling to understand it at a high-level. Can you include a comment describing what the condition(s) needing to be satisfied are that determine whether "trying to pack" succeeds or not?

Reply:
I tried to describe it in detail, but not sure it is easier to understand. Please let me know if it is still not clear.